### PR TITLE
Cleanup some issues

### PR DESCRIPTION
**Pull request by tbeu**

### CODE REVIEW

File: src/canonmn.cpp [hunk 1]
```diff
@@ -1819,13 +1819,15 @@ namespace Exiv2 {
...
-            float fu = pos->value().toFloat(2);
-            if (fu != 0.0) {
-                ltfl.focalLengthMin_ = pos->value().toLong(1) / fu;
-                ltfl.focalLengthMax_ = pos->value().toLong(0) / fu;
+        if (pos != metadata->end()) {
+            const Value &value = pos->value();
+            if (   value.count() >= 3
+                && value.typeId() == unsignedShort) {
+                float fu = value.toFloat(2);
+                if (fu != 0.0) {
```
Comment (D4N) [inline src/canonmn.cpp]: `0.0` is a double constant, please make it a `0.0f`

**Comment by D4N** [comment]
It looks good to me, but I am unsure about some of the changes and would like if someone else also took a look.

**Comment by clanmills** [comment]
I’ll review this later today.  Thomas: can you explain the rationale for those changes?

**Comment by tbeu** [comment]
My motivation was the code-analysis tool I am running and to learn how to contribute (by PRs as I see),

**Comment by piponazo** [comment]
I do not see any risk with the changes.

It is interesting that you mentioned the PVS-Studio tool. I am thinking about running some kind of automatic checks with that tool in the Exiv2 project, since the tool can be used freely in Open Source projects :)

**Review by piponazo — APPROVED** [review]

**Comment by D4N** [comment]
Uhm, I just took a look at the `stringFormat` function that @tbeu modified and I think that function is not correct (the original, not @tbeu's fault).

If I understand it correctly, it tries to allocate a large enough buffer to write the formatted string into it (and if it was not large enough allocate a larger buffer). But I think the detection when the buffer is large enough is wrong. According to the man pages `vsnprintf` returns the number of bytes that would be needed to save the formatted string regardless whether it actually could write all of them (provided that no error occurred). But we use `rc > 0` as an end condition for the loop, which it shouldn't be because `vsnprintf(NULL, 0, format, args)` already returns the required number of bytes to write the formatted string although it does not write anything. If you want to check this yourself, simply set the initial guess for `need` to `strlen(format)` (instead of `strlen(format)*2`) and run the test suite.

I think this function should at first create an std::string with the initial guess for `need` as the allocated length and run `vsnprintf`. Then we know from the return value if the initial guess was good (-> we are done) or not -> resize the buffer and call vsnprintf a second time.

**Comment by piponazo** [comment]
Good catch @D4N . I think this might be a perfect candidate to write some unit tests with google tests :)

**Comment by tbeu** [comment]
Do you want me to fix stringFormat? Can do it?

**Comment by D4N** [comment]
If you could fix the function, that would be terrific. Can you also add
a unit test (there are some in the unitTests directory)?

**Comment by piponazo** [comment]
I just noticed that this function is a private header of the library and I think that the unit tests infrastructure is not ready to tests the internals of the library. I will try to fix that situation so we can also test the internal headers.

**Comment by clanmills** [comment]
Thank You, Gentlemen.  This is great.  I go to bed in California and when I get up in the morning, you've fixed the code.  Very good work, Thank You Very Much.

I'd like to call a Team Meeting when I get home.  Sunday November 12 14:00 UTC.  I hope you can join us.  I'll send the agenda (and discussion slides) next week.

**Comment by tbeu** [comment]
OK, fixed stringFormat and tested manually. Feel free to add some proper unit tests by pushing to my branch some-cleanup.

**Comment by D4N** [comment]
I am against the changes you introduced in f980a2d. While `RafImage::printStructure` is only called with `option` being `kpsBasic` or `kpsRecursive`, there is nothing that prevents downstream users & developers to call this function with `option` set to something else. Imho the condition should stay, unless it does not make sense to use different values for `option` in this context (but then a warning/error should be added).

**Comment by tbeu** [comment]
Hm, f980a2de4778cd5a752f3a9b842ecf7818cc9b70 simply removes the nested checks for bPrint:
```c
if (bPrint) {
  if (bPrint) {
    // print here
  }
}
```

**Comment by D4N** [comment]
Whoops, my bad. You are of course right. Maybe we could even remove the scopes?

**Comment by tbeu** [comment]
Yes, no need to keep the scopes now.

File: include/exiv2/tiffimage.hpp [hunk 2]
```diff
@@ -90,7 +90,7 @@ namespace Exiv2 {
                 not valid (does not look like data of the specific image type).
           @warning This function is not thread safe and intended for exiv2 -p{S|R} as a file debugging aid
          */
-        virtual void printStructure(std::ostream& out, PrintStructureOption option,int depth=-1);
```
Comment (D4N) [inline include/exiv2/tiffimage.hpp]: The previous version sure looks odd. @clanmills Do you know why `depth` defaults to -1? You have added that parameter in fd5e983746c336336039e91cb6b656cf8eeccdea and set a default in the base class and here, so I suspect there is a reason for that.

File: src/tiffimage.cpp [hunk 3]
```diff
@@ -188,7 +188,7 @@ namespace Exiv2 {
         // recursively print the structure to /dev/null to ensure all metadata is in memory
         // must be recursive to handle NEFs which stores the raw image in a subIFDs
         std::ofstream devnull;
-        printStructure(devnull,kpsRecursive,0);
+        printStructure(devnull,kpsRecursive);
```
Comment (D4N) [inline src/tiffimage.cpp]: I'd rather keep it for consistency with the other `*image.cpp` files where depth is explicitly specified.

**Review by D4N — COMMENTED** [review]
Most of this looks good to me. I'd however like @clanmills to look at some of the changes (pinged him explicitly there) before we merge this.
Also, if you intend to submit further patches, could you make another PR out of that? This is already becoming quite huge.

**Comment by piponazo** [comment]
I tend to agree with @D4N . We could have merged the previous status of the PR, but now we need to review the new changes and it is more difficult to do it in the same PR since the number of changes is increasing.

Could you move the new commits to a new branch and create another PR ? In that way, we can review the previous changes (that were already reviewed).

Furthermore, after taking out the new commits from the branch, please rebase this branch on top of master (right now it is out of sync)

**Comment by tbeu** [comment]
OK, rebased and split the PRs.

**Comment by piponazo** [comment]
Thanks @tbeu. I am merging this now since the changes were approved before you introduced the additional changes. @D4N You made some comments that we could address separately. Do you think we should create other issue for those ?

**Pull request merged**